cura/Settings/ContainerManager.py

```diff
@@ -394,7 +583,34 @@ def _updateContainerNameFilters(self):
             name_filter = "{0} ({1})".format(mime_type.comment, suffix_list)
             self._container_name_filters[name_filter] = entry
 
-    # Factory function, used by QML
-    @staticmethod
-    def createContainerManager(engine, js_engine):
-        return ContainerManager()
+    def _getQualityContainers(self, quality_name, container_type = "quality_changes"):
+        global_stack = UM.Application.getInstance().getGlobalContainerStack()
+        if not global_stack:
+            return False
+
+        criteria = { "type": container_type, "name": quality_name }
+
+        filter_by_material = False
+
+        if global_stack.getMetaDataEntry("has_machine_quality"):
+            criteria["definition"] = global_stack.getBottom().getId()
+
+            filter_by_material = global_stack.getMetaDataEntry("has_materials")
+
+        material_ids = []
+        if filter_by_material:
+            stacks = [ s for s in cura.Settings.ExtruderManager.getInstance().getMachineExtruders(global_stack.getId()) ]
+            stacks.insert(0, global_stack)
+
+            for stack in stacks:
+                material_ids.append(stack.findContainer(type = "material").getId())
+
+        containers = UM.Settings.ContainerRegistry.getInstance().findInstanceContainers(**criteria)
+        for container in containers:
+            if filter_by_material and container.getMetaDataEntry("material") not in material_ids:
+                continue
```

Comment:
This at least warrants some documentation. Why do you only want profiles with a material that is currently in one of the extruder stacks?